keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -370,55 +399,23 @@ The leader election controller will then pick the leader based on its criteria f
-policy](https://kubernetes.io/releases/version-skew-policy/).
-
-Strategy is an option specified by a `LeaseCandidate` object, in most cases the
-controller responsible renewing the `LeaseCandidate` lease. The value for the
-strategy between different `LeaseCandidate` objects leading the same `Lease`
-should be the same, but during mixed version states, there is a possibility that
-they may differ. We will use a consensus protocol that favors the algorithm with
-the highest priority. The priority is a fixed list that is predetermined. For
-now, this is `NoCoordination` > `MinimumCompatibilityVersion`. For example, if
-three `LeaseCandidate` objects exist and two objects select
-`MinimumCompatibilityVersion` while the third selects `NoCoordination`,
-`NoCoordination` will take precedent and the coordinated leader election
-controller will use `NoCoordination` as the election strategy. The final
-strategy used will be written to the `Lease` object when the CLE controller
-creates the `Lease` for a suitable leader. This has the benefit of providing
-better debugging information and allows short circuiting of an election if the
-set of candidates and selected strategy is the same as before.
+There are cases where a user may want to change the leader election algorithm
+and this can be done via the `spec.Strategy` field in a Lease.
 
-#### Alternative for Strategy
-
-The current selected approach for handling `Strategy` has the complexity of components not having consensus and the CLE controller having to tiebreak decisions. This section highlights some alternatives
-
-Instead of having the LeaseCandidate controllers select the strategy, have the CLE controller directly select the strategy. This has the benefit that no client-go change is needed and bug fixes will land quicker via only kube-apiserver changes. Because the Lease object is created by the CLE controller, it must retrieve the default value for `Strategy` from somewhere. There are a couple other ways to populate the default value instead of via  the `LeaseCandidate`.
+The `Strategy` field signals to the coordinated leader election controller the
+appropriate algorithm to use when selecting leaders.
 
-1. Via writes to the `Strategy` field in Lease
-
-If the CLE controller always defaults to a value, a user may update the Lease
-object to take on a different strategy.
-
-```
-kind: Lease
-spec:
-  strategy: "MaximumCompatibilityVersion"
-```
+We will allow the Coordinated Leader Election controller to create a Lease
+without a holder. This will have a short duration of 5s. During this time, the
+`Lease` may be updated by a third party to the desired `spec.Strategy`. The
+strategy will always default to `MinimumCompatibilityVersion`, but may be
+changed during the initial step of a "leader-less" `Lease`. After the initial 5
+seconds, the coordinated leader election controller will proceed with the
+election using the data from `spec.Strategy`.
```

Comment:
This confuses me again. Why is the CLE special? Can't any party just set the stragegy, either through create or through update (like a human wanting to canary a component) ? Why such complicated 5s logic?

Reply:
I.e. these semantics: 
- if there are no candidate objects, `strategy` is empty and stays empty (legacy mode)
- if there are candidates, CLE will default the strategy to `MinimumCompatibilityVersion`.
- if there are candidates objects and they have opinions, change the strategy on create or via update. It will be sticky. CLE won't touch it.

Reply:
@sttts You're list looks right to me.  This is along the lines of what I was expecting.  The key thing I was hoping to see was support for both create and update in the third bullet point (exactly as you have it), so that it's clear that any authorized writer in the system has the opportunity to set or change the strategy when desired, including before the controller is started.

Reply:
I think that's possible, but I want to just justify the 5s logic a bit. 

>- if there are no candidate objects, strategy is empty and stays empty (legacy mode). Components do >election themselves.
>- if there are candidates, CLE will default the strategy to MinimumCompatibilityVersion.

The first two points are satisfied regardless of approach.

> if there are candidates objects and they have opinions (or a human has), make them (or the human) change the strategy on create or via update. It will be sticky. CLE won't touch it.

- To make it "sticky", we may need to rethink how releasing the lease works. Currently, releasing a lease involves deleting the lease object and letting leader election determine the next lease. The strategy information will disappear since the resource is deleted. To solve this, we can persist leases such that releasing does not involve deletion, but may run into things like unable to determine orphaned leases. Alternatively, would we require users to update the Lease every time the leader changes? In theory, a lot of user initiated strategy changes (like canary) are temporary, and it could be good to reset to default after a certain period. 
- candidates are not responsible for creating the Lease, the CLE is responsible. It doesn't make sense for candidates to create  the lease because they do not have the full view of other candidates, only the CLE controller maintains the full picture. Is this a point we need to revisit? Candidates will only be able to change the strategy via update, and CLE may have already gone through a round of leader selection.

Reply:
I understand the deletion problem. Clearly deleting the lease is an anti-pattern then.

I don't understand the last point. If there is a way to create a leaderless lease, and an actor has an opinion beyond the default behaviour, that's the way to state that.

> However, setting it is essentially a noop since CLE will not operate on Lease objects that have no candidates.

exactly. So won't the candidates beyond creating/updating the leaderless lease. They will just ensure their opinion is stated and CLE will pick it up when it is time.

Reply:
To be on the same page: how does a leaderless lease look like in detail? From the former discussion I assumed we have a canonical, validating shape for that.

Reply:
+1, let's define it clearly.  I was intuitively expecting it to look very close to an expired lease to minimize the number of distinct states..

Reply:
> Giving up leadership would mean to update to that leaderless state, instead of deletion.

ack, yes this part is what I imagined a leaderless lease would look like and solves the deletion problem. I think the shape of the leaderless lease is not a point of contention and is exactly as described (looks like an expired lease, has all the fields of a Lease except the holder and renewTime). The main thing we're trying to resolve is the creator of the leaderless lease. 

Correct me if I'm wrong but I think what you're saying is that a leaderless lease can be created by anyone (thus solving both create and update case for setting the strategy). I'm still under the impression that it may only be created by the CLE controller and feels like a mixup of responsibilities allowing for instance candidates to create a leaderless Lease that they do not claim. 

My 5s idea was based on the idea that if only the CLE controller was able to create an empty lease, then specifically for "strategy on create" cases, components need time some time to set the default before the CLE controller starts assuming a default strategy and selecting leaders. This does get resolved if an interested candidate chooses to create the leaderless lease themselves to set the strategy. I'm curious if that was the suggestion or if there's more regarding the point *"if there are candidates objects and they have opinions (or a human has), make them (or the human) change the strategy on create or via update. It will be sticky. CLE won't touch it."*

Reply:
5s or any other timing dependent mechanism feels like both a Kube anti-pattern and it seems to be brittle. 5s in pod startup time terms is super small.

Reply:
Spent some more time thinking about it, and yeah I think the approach that allows anyone to create an empty lease is sound. It should still be recommended to stick with the default "MinimumCompatibilityVersion" strategy, but if we already see that a lease has been created, the strategy used will be respected and used.